src/platform/datapath_raw_socket.c

```diff
@@ -585,6 +783,82 @@ CxPlatDpRawParseUdp(
     Packet->BufferLength = Length;
 }
 
+_IRQL_requires_max_(DISPATCH_LEVEL)
+static
+void
+CxPlatDpRawParseTcp(
+    _In_ const CXPLAT_DATAPATH* Datapath,
+    _Inout_ CXPLAT_RECV_DATA* Packet,
+    _In_reads_bytes_(Length)
+        const TCP_HEADER* Tcp,
+    _In_ uint16_t Length
+    )
+{
+    uint16_t HeaderLength;
+    if (Length < sizeof(TCP_HEADER)) {
+        QuicTraceEvent(
+            DatapathErrorStatus,
+            "[data][%p] ERROR, %u, %s.",
+            Datapath,
+            Length,
+            "packet is too small for a TCP header");
+        return;
+    }
+
+    HeaderLength = Tcp->HeaderLength * sizeof(uint32_t);
+    if (Length < HeaderLength) {
+        QuicTraceEvent(
+            DatapathErrorStatus,
+            "[data][%p] ERROR, %u, %s.",
+            Datapath,
+            Length,
+            "packet is too small for a TCP header");
+        return;
+    }
+
+    Length -= HeaderLength;
+
+    //
+    // We only handle 3 types of TCP packets:
+    // 1. Pure ACKs that carry at least one byte data.
+    // 2. SYNs and SYN+ACKs for TCP handshake.
+    // 3. FINs for graceful shutdown.
+    //
+    // Packets that don't match the rules above are discarded.
+    //
+    if (Tcp->Flags == TH_ACK && Length > 0) {
+        //
+        // Only data packets with only ACK flag set are indicated to QUIC core.
+        //
+        Packet->Reserved = L4_TYPE_TCP;
+        Packet->Route->TcpState.AckNumber = Tcp->SequenceNumber;
+        Packet->Route->TcpState.SequenceNumber = Tcp->AckNumber;
```

Comment:
It's a bit confusing that you are setting the route to the inverse of the TCP header. Perhaps a comment explaining why?

Reply:
Reversed the order here and added a comment in RouteUpdate.